### DISCUSSION TITLE
How to add custom keymap group for those keymaps related to the default "fold" operations in LazyVim?

### DISCUSSION THREAD

**Discussion by hopezh** [LazyVim/LazyVim#788] [Q&A] [answered]
I noticed that LazyVim added a group `+goto` for those default keymaps related to operations starting with `g`, as shown below.
![image](https://github.com/LazyVim/LazyVim/assets/23435810/4ccec1aa-5fe7-4b62-b640-f5e89a34491c)

So, I want to add a keymap group `fold` for those related to the default fold operations starting with `z`, as shown below. 
![image](https://github.com/LazyVim/LazyVim/assets/23435810/f488eef8-56a1-4336-825a-8ef244cfe237)

I added the following code in `.config/LazyVim/lua/plugins/which_key.lua`. However, this seems to wash away all the default which-key groups and descriptions, and I can only delete this `which_key.lua` file in the `plugins` folder to get the default keymap settings back. 
![image](https://github.com/LazyVim/LazyVim/assets/23435810/1f6d4821-2db9-462f-a488-302fa7fbc01f)

My question is: **_How to add custom keymap and keymap group for those that are not associated with a particular plugins, especially those that dosen't use a leader key, without affecting the default settings of `which-key` that come with `LazyVim` ?_** 

Thanks.

**Comment by folke** [comment, marked as answer]
There's an easier way in LazyVim:

```lua
{
    "folke/which-key.nvim",
    opts = {
      defaults = {
        ["z"] = { name = "+fold" }
      }
    }
}
```

**Reply by hopezh** [reply]
Thanks, it works. 
![image](https://github.com/LazyVim/LazyVim/assets/23435810/f90e1503-cc40-4353-82ce-376588ebc8b9)

However, when I added the following code to add two custom keymaps and a keymap group in the `.config/lazyvim/lua/plugins/which_key.lua` file, the default which-key settings are washed away, again.
![image](https://github.com/LazyVim/LazyVim/assets/23435810/4b9e2c4b-f035-4861-afd7-48cefa9a0a08)

May I ask how to add a custom keymap for a plugin inside this which-key config file, rather than in the config file of the plugin itself? 

Thanks.

**Reply by hopezh** [reply]
@folke I marked your reply above as the answer to my original quesiont. 

But I hope you can kindly advise on the issue I mentioned above after your reply. 

Thank you.

**Reply by folke** [reply]
That's because you're completely overriding the LazyVim options.
See https://github.com/LazyVim/LazyVim/issues/342#issuecomment-1445929800 on how to do this

**Reply by hopezh** [reply]
Thanks for the advice. 

After a few rounds of trial-and-errors, I managed to get a config that works: (I didn't know that we can use `require("which-key").register({})` in a function twice...) 

```lua
return {
    "folke/which-key.nvim",

    opts = function()
        -- add keymap group for those keybindings not using a leader key
        require("which-key").register({   
            ["z"] = { name = "+fold" },
        })

        -- add keymaps and groups for those keybindings using a leader key
        require("which-key").register({  
            z = {
                name = "OatKey",
                m = { "<cmd>CellularAutomaton make_it_rain<CR>", "Make it rain..." },
                g = { "<cmd>CellularAutomaton game_of_life<CR>", "Game of life..." },
            },
        }, { prefix = "<leader>" })
    end,
}

```

**Reply by dpetka2001** [reply]
You don't actually need to call it twice. You can also do it all in one like this
```lua
    opts = function()
      require("which-key").register({
        ["z"] = { name = "+fold" },
        ["<leader>z"] = {
          name = "+OatKey",
          m = { "<cmd>CellularAutomaton make_it_rain<CR>", "Make it rain..." },
          g = { "<cmd>CellularAutomaton game_of_life<CR>", "Game of life..." },
        },
      })
    end,
```
There are other examples in the repo of `which-key` which can help you with other combinations.

**Reply by hopezh** [reply]
Thanks a lot, and I learned more about LazyVim.
